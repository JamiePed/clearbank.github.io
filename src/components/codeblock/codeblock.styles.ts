import styled from 'styled-components'
import { colors } from 'src/components/theme'

export const Container = styled.article`
  position: relative;
  margin-bottom: 30px;
  border-radius: 5px;
  overflow: hidden; // needed for border-radius on parent
`

export const Header = styled.header`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 13px 5px 15px 15px;
  background-color: ${({ color }) => color};
  font-size: 16px;
  color: ${colors.brandLight};
`

export const RequestDetails = styled.h4`
  margin: 0;
  padding-right: 5px;
  font-size: 16px;
  line-height: 20px;
  color: inherit;
  text-transform: capitalize;
  font-family: 'Roboto Mono', monospace;
`
export const CopyButton = styled.button.attrs({
  type: 'button'
})`
  padding: 0;
  border: none;
  font-family: inherit;
  line-height: 1.15;
  font-size: 13px;
  color: inherit;
  background: none;
  cursor: pointer;
`

export const Body = styled.div`
  padding: 20px 2px 20px 15px;
  font-family: 'RobotoMono-Regular', 'Roboto Mono', sans-serif;
  font-size: 14px;
  font-weight: 400;
  font-style: normal;
  text-align: left;
  line-height: normal;
  background: #1e1e1e;
  overflow: visible;
  overflow-x: scroll;
`

export const Overlay = styled.div`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  pointer-events: none;
`
