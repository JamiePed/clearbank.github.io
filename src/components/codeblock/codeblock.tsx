import React, { useState, useEffect, useRef } from 'react'
import CopyToClipboard from 'react-copy-to-clipboard'

import CodeHighlight from 'src/components/code-highlight'
import CodeOverlay from 'src/components/code-overlay'

import * as Styles from './codeblock.styles'
import * as Types from './codeblock.types'

const Codeblock: React.FunctionComponent<Types.CodeblockProps> = ({
  codeblock
}: Types.CodeblockProps) => {
  const { title, codeSnippet, copyCode, color, language } = codeblock
  const [showCodeOverlay, setShowCodeOverlay] = useState(false)
  let codeOverlayTimeoutId = -1

  useEffect(() => {
    // eslint-disable-next-line react-hooks/exhaustive-deps
    codeOverlayTimeoutId = setTimeout(() => setShowCodeOverlay(false), 2500)

    return () => {
      clearTimeout(codeOverlayTimeoutId)
    }
  }, [showCodeOverlay])

  return (
    <Styles.Container>
      <Styles.Header color={color}>
        <Styles.RequestDetails>{title}</Styles.RequestDetails>
        {copyCode && (
          <CopyToClipboard
            text={codeSnippet}
            onCopy={() => setShowCodeOverlay(!showCodeOverlay)}
          >
            <Styles.CopyButton>Copy code</Styles.CopyButton>
          </CopyToClipboard>
        )}
      </Styles.Header>
      <Styles.Body>
        <CodeHighlight code={codeSnippet} language={language} />
      </Styles.Body>
      <Styles.Overlay>
        <CodeOverlay showCodeOverlay={showCodeOverlay} message='Code copied' />
      </Styles.Overlay>
    </Styles.Container>
  )
}

export default Codeblock
