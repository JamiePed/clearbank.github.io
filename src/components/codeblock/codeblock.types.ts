import { CodeblockType } from 'src/global.types'

export interface CodeblockProps {
  codeblock: CodeblockType
}
