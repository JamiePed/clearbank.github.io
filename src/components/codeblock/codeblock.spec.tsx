import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'

import Component from './codeblock'

import * as Types from './codeblock.types'

let component: ReturnType<typeof render> = null
let fragment: DocumentFragment = null
let root: HTMLElement = null

const defaultProps: Types.CodeblockProps = {
  codeblock: {
    title: 'title',
    codeSnippet: 'hello-world',
    section: 'request',
    language: 'turbo-pascal',
    copyCode: false,
    color: 'red'
  }
}

beforeEach(() => {
  component = render(<Component {...defaultProps} children='children' />)

  fragment = component.asFragment()
  root = component.container.firstChild
})

afterEach(() => {
  fragment = null
  root = null
})

describe('Component', () => {
  test('matches snapshot', () => {
    expect(root).toMatchSnapshot()
  })

  test('has child elements', () => {
    expect(root.querySelectorAll('header').length).toBeGreaterThanOrEqual(1)
    expect(root.querySelectorAll('header h4').length).toBeGreaterThanOrEqual(1)
    expect(root.querySelectorAll('pre').length).toBeGreaterThanOrEqual(1)
  })

  test('has title from props', () => {
    expect(root.textContent.includes(defaultProps.codeblock.title)).toBeTruthy()
  })

  test('has code-snippets from props', () => {
    expect(
      root.textContent.includes(defaultProps.codeblock.codeSnippet)
    ).toBeTruthy()
  })

  test('has language attributes from props', () => {
    expect(
      root.querySelectorAll(`[class*="${defaultProps.codeblock.language}"]`)
        .length
    ).toBeTruthy
  })

  test('has color attribute from props', () => {
    expect(root.querySelector('header')).toHaveAttribute('color')
  })

  test('is not empty', () => {
    expect(root.textContent.length).toBeTruthy()
  })
})
