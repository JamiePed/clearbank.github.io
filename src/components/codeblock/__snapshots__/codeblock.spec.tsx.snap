// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component matches snapshot 1`] = `
.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
}

.c6 {
  padding: 22px 42px;
  font-family: 'Roboto Mono',monospace;
  font-size: 13px;
  font-weight: bold;
  background-color: #f2f2f2;
  opacity: 0;
  -webkit-transition: opacity 1s;
  transition: opacity 1s;
}

.c0 {
  position: relative;
  margin-bottom: 30px;
  border-radius: 5px;
  overflow: hidden;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 13px 5px 15px 15px;
  background-color: red;
  font-size: 16px;
  color: #fff;
}

.c2 {
  margin: 0;
  padding-right: 5px;
  font-size: 16px;
  line-height: 20px;
  color: inherit;
  text-transform: capitalize;
  font-family: 'Roboto Mono',monospace;
}

.c3 {
  padding: 20px 2px 20px 15px;
  font-family: 'RobotoMono-Regular','Roboto Mono',sans-serif;
  font-size: 14px;
  font-weight: 400;
  font-style: normal;
  text-align: left;
  line-height: normal;
  background: #1e1e1e;
  overflow: visible;
  overflow-x: scroll;
}

.c4 {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  pointer-events: none;
}

<article
  class="c0"
>
  <header
    class="c1"
    color="red"
  >
    <h4
      class="c2"
    >
      title
    </h4>
  </header>
  <div
    class="c3"
  >
    <pre
      class="prism-code language-turbo-pascal"
      style="color: rgb(156, 220, 254); background-color: rgb(30, 30, 30);"
    >
      <div
        class="token-line"
        style="color: rgb(156, 220, 254);"
      >
        <span
          class="token plain"
        >
          hello-world
        </span>
      </div>
    </pre>
  </div>
  <div
    class="c4"
  >
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        Code copied
      </div>
    </div>
  </div>
</article>
`;
